components/collector/src/source_collectors/gitlab/base.py

```diff
@@ -139,53 +138,58 @@ def _lookback_datetime(self) -> datetime:
 
 
 @dataclass
-class Pipeline:
+class GitLabJSON:
+    """Base class for GitLab response JSON."""
+
+    @classmethod
+    def from_json(cls, **kwargs) -> Self:
+        """Override to ignore unknown fields so the caller does not need to weed the GitLab JSON."""
+        field_names = [field.name for field in fields(cls)]
+        return cls(**{key: value for key, value in kwargs.items() if key in field_names})
+
+
+@dataclass
+class Pipeline(GitLabJSON):
     """GitLab pipeline JSON. See https://docs.gitlab.com/ee/api/pipelines.html."""
 
     id: int
     project_id: int
+    name: str
     ref: str
     status: str
     source: str
     created_at: str
-    updated_at: str
     web_url: str
-    started_at: str | None = None
-    finished_at: str | None = None
-    duration: int | None = None  # Seconds
-
-    def __init__(self, **kwargs) -> None:
-        """Override to ignore unknown fields so the caller does not need to weed the GitLab pipeline JSON."""
-        field_names = [field.name for field in fields(self)]
-        for key, value in kwargs.items():
-            if key in field_names:
-                setattr(self, key, value)
+    updated_at: str = ""
+    schedule_description: str = ""  # Pipeline schedule description for scheduled pipelines
 
     @property
     def start(self) -> datetime:
-        """Return the pipeline start time.
-
-        The started_at field may be empty, be prepared. See https://gitlab.com/gitlab-org/gitlab/-/issues/210353.
-        """
-        return parse_datetime(self.started_at or self.created_at)
+        """Return the pipeline start time."""
+        return parse_datetime(self.created_at)
 
     @property
     def end(self) -> datetime:
-        """Return the pipeline end time.
-
-        The finished_at field may empty if the pipeline hasn't finished yet. Use the current time as fallback.
-        """
-        return parse_datetime(self.finished_at) if self.finished_at else datetime.now(tz=tzutc())
+        """Return the pipeline end time."""
+        return parse_datetime(self.updated_at) if self.updated_at else datetime.now(tz=tzutc())
 
     @property
     def datetime(self) -> datetime:
         """Return the datetime of the pipeline."""
         return parse_datetime(self.updated_at or self.created_at)
 
     @property
-    def pipeline_duration(self) -> timedelta:
+    def duration(self) -> timedelta:
         """Return the duration of the pipeline."""
-        return self.end - self.start if self.duration is None else timedelta(seconds=self.duration)
```

Comment:
Where did the old `self.duration` come from, was it something included in the json originally?

Reply:
In the previous version, Quality-time would get more detailed information about the pipeline with the `started_at`, `finished_at` and `duration` fields. However, this needs a call to the `pipeline/id` endpoint for each pipeline individually. Also, there is this open bug: https://gitlab.com/gitlab-org/gitlab/-/issues/19594. So I opted to simply used the `created_at` and `updated_at` fields as proxy for the pipeline duration.